### PR TITLE
server: enhance comments surrounding draining

### PR DESCRIPTION
**Pull request by cameronnunez**
This patch improves the comments related to the draining process,
fixing incorrect statements and providing more details to make
the process more easily understood.

Release note: None

### CODE REVIEW

**Comment by cameronnunez** [comment]
@ZhouXing19 we can be a bit more precise about how we think about the drain sequence. While we have been thinking about as:
1. Initial wait
2. Wait for SQL connections to close
3. Drain SQL connections
4. Drain leases & leadership

Really it's better to think about phase 3 as: **Drain the SQL layer**. This includes draining SQL connections, draining distributed SQL execution flows, and draining SQL table leases.

**Comment by cameronnunez** [comment]
cc @taroface

File: docs/generated/settings/settings-for-tenants.txt [hunk 1]
```diff
@@ -60,9 +60,9 @@ server.oidc_authentication.provider_url	string		sets OIDC provider URL ({provide
...
-server.shutdown.lease_transfer_wait	duration	5s	the amount of time a server waits to transfer range leases before proceeding with the rest of the shutdown process (note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
-server.shutdown.query_wait	duration	10s	the server will wait for at least this amount of time for active queries to finish (note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
+server.shutdown.drain_wait	duration	0s	the minimum amount of time a server waits in an unready state before proceeding with a drain(note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
+server.shutdown.lease_transfer_wait	duration	5s	the timeout for a single iteration of the range lease transfer phase of draining(note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
```
Comment (taroface) [inline docs/generated/settings/settings-for-tenants.txt:64]: Nit: missing space between "drain" and "(note"

**Review by knz — COMMENTED** [review]
nice, thanks

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-:-MvP0EO26f5fAwFKE4pM:btjme7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cameronnunez and @ZhouXing19)

___
*[docs/generated/settings/settings-for-tenants.txt, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-nTw6J97TqF98gf3:-MvP-nTw6J97TqF98gf4:b-rxq7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/docs/generated/settings/settings-for-tenants.txt#L63)):*
> ```help
> server.oidc_authentication.scopes	string	openid	sets OIDC scopes to include with authentication request (space delimited list of strings, required to start with `openid`)
> server.rangelog.ttl	duration	720h0m0s	if nonzero, range log entries older than this duration are deleted every 10m0s. Should not be lowered below 24 hours.
> server.shutdown.drain_wait	duration	0s	the minimum amount of time a server waits in an unready state before proceeding with a drain(note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
> ```

why "minimum" here? Is there a way for the initial delay to be longer than the configured value?
___
*[pkg/kv/kvserver/store.go, line 1358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-qNY0ilwvkYAnNyx:-MvP-qNY0ilwvkYAnNyy:bhpzu35) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/pkg/kv/kvserver/store.go#L1358)):*
> ```Go
> //
> // Note: this code represents ONE round of draining. This code is iterated on
> // indefinitely until all leases are transferred away.
> ```

perhaps you can point the reader to where this iteration can be found.
___
*[pkg/server/drain.go, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-zQT0BeVC2qZgrRi:-MvP-zQT0BeVC2qZgrRj:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/pkg/server/drain.go#L174)):*
> ```Go
> // Drain idempotently activates the draining mode.
> // Note: this represents ONE round of draining. This code is iterated on
> // indefinitely until all range leases have been drained.
> ```

ditto
___
*[pkg/server/drain.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP00vo0aN4H5JUsQJq:-MvP00vo0aN4H5JUsQJr:b-vyajxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc1957a534ca579e0ad0d155440f0278dbc1af2d/pkg/server/drain.go#L230)):*
> ```Go
> 		return err
> 	}
> 	// Finally, mark the node as draining in liveness and drain the
> ```

We need to explain the bit about liveness too.

**Review by ZhouXing19 — COMMENTED** [review]
:lgtm:! My opinions are just for minor changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-:-MvPAUdbCDLZMBxevAT3:b-2eus6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cameronnunez, @knz, and @taroface)

___
*[docs/generated/settings/settings-for-tenants.txt, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-nTw6J97TqF98gf3:-MvP1Vlq-DhndFpQDBpP:b-bnekeg) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/docs/generated/settings/settings-for-tenants.txt#L63)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

why "minimum" here? Is there a way for the initial delay to be longer than the configured value?
</blockquote></details>

I think it was trying to say that this is a "hard wait" without any early exit.
___
*[docs/generated/settings/settings-for-tenants.txt, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvK7qny9AxPNRUw3mGP-r1-63:-MvP3GWW5xdkNI2N6YRt:bvjyq0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/docs/generated/settings/settings-for-tenants.txt#L63)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

My understanding is that the unready state is the first step of a drain. So should "before proceeding with a drain" be e.g. "at the beginning of a drain"?
</blockquote></details>

It seems that we try to say this period is not a draining phase, but more of a "warm-up" that waits for the draining to starts. 
___
*[pkg/server/drain.go, line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP9Pvx7gqVxlMO9Jxi:-MvP9Pvx7gqVxlMO9Jxj:b-wt65n0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/pkg/server/drain.go#L248)):*
> ```Go
> 
> 	// Set the gRPC mode of the node to "draining" and mark the node as "not ready".
> 	// Probes to /health?ready=1 will now notice the change in the node's readiness.
> ```

nit: "will now notice the node is in a "shutting-down" status, and is unready for new SQL connection."
___
*[pkg/server/drain.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP9kUG5VZneFdZ2BXU:-MvP9kUH3oSZT90SKJYY:b-n2dzp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/pkg/server/drain.go#L261)):*
> ```Go
> 
> 	// Drain all SQL connections.
> 	// The queryWait duration is used to wait on clients to self-disconnect.
> ```

Nit: Though this may have been conveyed by the words "wait" and "timeout", maybe we can explicitly add that "Once all existing SQL queries finish before the `queryWait` timeout, the server early exits this waiting phase and proceeds to the rest of the drain".

File: docs/generated/settings/settings-for-tenants.txt [hunk 2]
```diff
@@ -60,9 +60,9 @@ server.oidc_authentication.provider_url	string		sets OIDC provider URL ({provide
...
 server.rangelog.ttl	duration	720h0m0s	if nonzero, range log entries older than this duration are deleted every 10m0s. Should not be lowered below 24 hours.
-server.shutdown.drain_wait	duration	0s	the amount of time a server waits in an unready state before proceeding with the rest of the shutdown process (note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
-server.shutdown.lease_transfer_wait	duration	5s	the amount of time a server waits to transfer range leases before proceeding with the rest of the shutdown process (note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
-server.shutdown.query_wait	duration	10s	the server will wait for at least this amount of time for active queries to finish (note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
+server.shutdown.drain_wait	duration	0s	the minimum amount of time a server waits in an unready state before proceeding with a drain(note that the --drain-wait parameter for cockroach node drain may need adjustment after changing this setting)
```
[the remark this answers is not in the document]
Comment (cameronnunez) [reply docs/generated/settings/settings-for-tenants.txt:63]: The purpose of the first step is to wait the duration of `drainWait` before we start draining. So while it is technically a part of the drain sequence, it is not a part of the actual drain.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76172)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cameronnunez, @taroface, and @ZhouXing19)

___
*[docs/generated/settings/settings-for-tenants.txt, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-nTw6J97TqF98gf3:-MvQr98dCu6rZ80snVOI:bhirgab) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/docs/generated/settings/settings-for-tenants.txt#L63)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

I think it was trying to say that this is a "hard wait" without any early exit.
</blockquote></details>

then the good word is "fixed", not "minimum".

**Review by cameronnunez — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @taroface, and @ZhouXing19)

___
*[docs/generated/settings/settings-for-tenants.txt, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-nTw6J97TqF98gf3:-MvUPUaqBhlFgiisKD4T:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/docs/generated/settings/settings-for-tenants.txt#L63)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

then the good word is "fixed", not "minimum".
</blockquote></details>

Fixed.
___
*[docs/generated/settings/settings-for-tenants.txt, line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvK7qny9AxPNRUw3mGP-r1-64:-MvUPXCTB86iQi6CGITZ:brvdsst) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/docs/generated/settings/settings-for-tenants.txt#L64)):*
<details><summary><i>Previously, taroface (Ryan Kuo) wrote…</i></summary><blockquote>

Nit: missing space between "drain"/"draining" and "(note" in the two above settings
</blockquote></details>

Noted fixed.
___
*[pkg/kv/kvserver/store.go, line 1358 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-qNY0ilwvkYAnNyx:-MvUP_-33aElwKe1E3mT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/pkg/kv/kvserver/store.go#L1358)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

perhaps you can point the reader to where this iteration can be found.
</blockquote></details>

Done.
___
*[pkg/server/drain.go, line 174 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-zQT0BeVC2qZgrRi:-MvUPaJb5-AvjN5enMF4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/pkg/server/drain.go#L174)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.
___
*[pkg/server/drain.go, line 230 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP00vo0aN4H5JUsQJq:-MvUPby09wZ6l4CNBjdH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc1957a534ca579e0ad0d155440f0278dbc1af2d/pkg/server/drain.go#L230)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

We need to explain the bit about liveness too.
</blockquote></details>

Done.
___
*[pkg/server/drain.go, line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP9Pvx7gqVxlMO9Jxi:-MvUPcagA9In5UoJvUDA:b-us3qyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/pkg/server/drain.go#L248)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

nit: "will now notice the node is in a "shutting-down" status, and is unready for new SQL connection."
</blockquote></details>

since the mode is `modeDraining` I think saying "draining" is more appropriate. Also the node can still accept new SQL connections here.
___
*[pkg/server/drain.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP9kUG5VZneFdZ2BXU:-MvUPpro0IGQkClsGpTf:ba3zv4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/pkg/server/drain.go#L261)):*
<details><summary><i>Previously, ZhouXing19 (Jane Xing) wrote…</i></summary><blockquote>

Nit: Though this may have been conveyed by the words "wait" and "timeout", maybe we can explicitly add that "Once all existing SQL queries finish before the `queryWait` timeout, the server early exits this waiting phase and proceeds to the rest of the drain". 
</blockquote></details>

More specific about it being a timeout; timeout implies a max duration.

File: docs/generated/settings/settings-for-tenants.txt (hunk 1 above)
Comment (cameronnunez) [reply docs/generated/settings/settings-for-tenants.txt:64]: Fixed thanks.

**Review by knz — COMMENTED** [review]
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cameronnunez, @taroface, and @ZhouXing19)

___
*[docs/generated/settings/settings-for-tenants.txt, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-nTw6J97TqF98gf3:-MvUQB2L-UJOorRv_f4G:bvwi5fy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/docs/generated/settings/settings-for-tenants.txt#L63)):*
<details><summary><i>Previously, cameronnunez (Cameron Nunez) wrote…</i></summary><blockquote>

Fixed.
</blockquote></details>

i don't see it

also there's a space missing before a parenthesis

**Review by cameronnunez — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz, @taroface, and @ZhouXing19)

___
*[docs/generated/settings/settings-for-tenants.txt, line 63 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76172#-MvP-nTw6J97TqF98gf3:-MvVerq3AFo4O26c7PIJ:b-qeewvw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f69fee2e1a7fe1a56018579df68ec6b876bc0f90/docs/generated/settings/settings-for-tenants.txt#L63)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

i don't see it

also there's a space missing before a parenthesis
</blockquote></details>

ah I was having issues with docs gen. Fixed.

**Review by knz — APPROVED** [review]
> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76172)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @taroface and @ZhouXing19)

**Comment by ajwerner** [comment]
bors r-

CI failure

**Comment by irfansharif** [comment]
PS, I think this is tripping up the merge queue due due to unexpected diffs: https://teamcity.cockroachdb.com/viewLog.html?buildId=4372990&problemId=2701

Have you tried `dev generate bazel` to see if it suggests changes?

**Comment by cameronnunez** [comment]
yeah I've noticed but the regeneration of that file seems odd to me, but I'll add the lines and see what happens with CI

**Comment by cameronnunez** [comment]
looks like I'm running into problems with `node_modules` because I switched between `make build` and `dev build`

**Comment by cameronnunez** [comment]
bors=knz

**Comment by cameronnunez** [comment]
bors r=knz

**Pull request merged**